tilavarauspalvelu/api/graphql/types/recurring_reservation/serializers.py

```diff
@@ -353,6 +355,80 @@ def save(self, **kwargs: Any) -> RecurringReservation:
         return instance
 
 
+class ReservationSeriesAddReservationSerializer(NestingModelSerializer):
+    """Add a reservation to a recurring reservation."""
+
+    instance: RecurringReservation
+
+    pk = serializers.IntegerField(required=True)
+
+    begin = serializers.DateTimeField(required=True, write_only=True)
+    end = serializers.DateTimeField(required=True, write_only=True)
+
+    buffer_time_before = serializers.DurationField(required=False, write_only=True)
+    buffer_time_after = serializers.DurationField(required=False, write_only=True)
+
+    class Meta:
+        model = RecurringReservation
+        fields = [
+            "pk",
+            "begin",
+            "end",
+            "buffer_time_before",
+            "buffer_time_after",
+        ]
+
+    def validate(self, data: dict[str, Any]) -> dict[str, Any]:
+        self.instance.validator.validate_has_reservations()
+
+        reservation_unit = self.instance.reservation_unit
+
+        begin: datetime.datetime = data["begin"].astimezone(DEFAULT_TIMEZONE)
+        end: datetime.datetime = data["end"].astimezone(DEFAULT_TIMEZONE)
+
+        if reservation_unit.reservation_block_whole_day:
+            data["buffer_time_before"] = reservation_unit.actions.get_actual_before_buffer(begin)
+            data["buffer_time_after"] = reservation_unit.actions.get_actual_after_buffer(end)
+
+        buffer_time_before = data.setdefault("buffer_time_before", reservation_unit.buffer_time_before)
+        buffer_time_after = data.setdefault("buffer_time_after", reservation_unit.buffer_time_after)
+
+        reservation_unit.validator.validate_begin_before_end(begin, end)
+        reservation_unit.validator.validate_reservation_begin_time_staff(begin=begin)
+        reservation_unit.validator.validate_no_overlapping_reservations(
+            begin=begin,
+            end=end,
+            new_buffer_time_before=buffer_time_before,
+            new_buffer_time_after=buffer_time_after,
+        )
+
+        return data
+
+    def update(self, instance: RecurringReservation, validated_data: dict[str, Any]) -> RecurringReservation:
+        now = local_datetime()
+
+        # Copy first reservation in the series as the new reservation.
+        # This retains data from many-to-one relationships, but not for many-to-many or one-to-many relationships.
+        # One-to-one relationships to or from the model would cause an error.
+        reservation: Reservation = instance.reservations.first()
+        reservation.id = None  # Create a new reservation
+        reservation.begin = validated_data["begin"]
+        reservation.end = validated_data["end"]
+        reservation.buffer_time_before = validated_data["buffer_time_before"]
```

Comment:
Non-zero buffers only if other than `ReservationTypeChoice.BLOCKED`

Reply:
Fixed logic in validation.